tracking-vuln/README.md

```diff
@@ -0,0 +1,78 @@
+
+### egg.js of the vulnerable extension
+The first thing we have to do is to replace the way we are getting the image; when directly setting the `src` attribute in an image tag we have no control over the outgoing HTTP-request. Thus we remove line 38. We also replace line 66 (which would add the egg to the page) with the following:
+```javascript
+fetch(imageSrc + color + "_egg.png", {referrerPolicy:"no-referrer"})
+	.then(res => res.blob())
+	.then(blob => {
+		imgSrc = URL.createObjectURL(blob)
+		egg.src = imgSrc
+		document.body.prepend(egg);
+	})
+``` 
+This code makes an HTTP-request to the image URL and specifies that we do not want to include the "referer" header. Then it transforms the response so that we get a URL that we can use for source for our image. Lastly we add the egg to the page.
+
+However, this introduces a new problem. When sending our HTTP-request in this fashion we get a new header in the request: "origin". This header will, similarly to the "referer", give away our visited website. To fix this we have to send our request through a proxy. We have provided a simple proxy server in the folder *proxy-server*, see the [section below](#proxyServer) for details. 
+
+We create a variable for the proxyServer address at the top of the file `let proxyServer = "the_proxy_address"`, and update the code snippet above to:
+```javascript
+fetch(proxyServer + "?url=" + imageSrc + color + "_egg.png", {referrerPolicy:"no-referrer"})
+	.then(res => res.blob())
+	.then(blob => {
+		imgSrc = URL.createObjectURL(blob)
+		egg.src = imgSrc
+		document.body.prepend(egg);
+	})
+```
+Here we send the image URL as a query parameter to our proxy.
+
+### app.js of the image-provider
+While this server is supposedly out of your control (hence why it being able to track you is a problem) it is worth noting that due to the change of how we perform the HTTP-request, it is necessary that the server allows [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) for your origin for any image you request. This would look something like this in the server code: `res.header("Access-Control-Allow-Origin", "*");`.
+
+### <a name="proxyServer">Proxy server</a>
+In the *proxy-server* folder, the `app.js` file contains a simple proxy server. This proxy server will serve the purpose of hiding your "origin" by being the one issuing the HTTP-request to the image provider. All it does is receiving your original HTTP-request and then redirecting it to the image provider.
+
+ To run it enter the folder and run `npm install` and then `node app.js`. The default port is 3001; this can be changed in the `app.js` file. Similarly to the malicious webserver (the image provider) this proxy needs to be made visible (for example with *ngrok*), and then have the URL it is visible on added to the `proxyServer` variable in the `egg.js` file.
```

Comment:
```suggestion
 To run it enter the folder and run `npm install` and then `node app.js`. The default port is 3001; this can be changed in the `app.js` file. Similarly to the malicious webserver (the image provider), this proxy needs to be made visible (for example with *ngrok*), and then have the URL it is visible on added to the `proxyServer` variable in the `egg.js` file.
```